File: packages/capacitor-plugin/android/src/main/java/com/otakit/updater/DocumentReadyBridge.java
package com.otakit.updater;

import android.net.Uri;
import android.webkit.WebView;
import androidx.webkit.ScriptHandler;
import androidx.webkit.WebViewCompat;
import androidx.webkit.WebViewFeature;
import java.util.Collections;
import org.json.JSONObject;

/** Captures the activation in the document's bridge before application JavaScript runs. */
final class DocumentReadyBridge {

  private ScriptHandler script;

  static void requireSupported() {
    if (!WebViewFeature.isFeatureSupported(WebViewFeature.DOCUMENT_START_SCRIPT)) {
      throw new IllegalStateException(
        "Update Android System WebView before applying OtaKit updates: document-start scripts are unavailable"
      );
    }
  }

  private static String origin(String appUrl) {
    return Uri.parse(appUrl).buildUpon().path(null).fragment(null).clearQuery().build().toString();
  }

  static void preflight(WebView webView, String appUrl) {
    requireSupported();
    // Validate this WebView and origin before publishing a trial in persistent state.
    WebViewCompat.addDocumentStartJavaScript(
      webView,
      "",
      Collections.singleton(origin(appUrl))
    ).remove();
  }

  static String source(String activationId) {
    return String.format(
      java.util.Locale.ROOT,
      """
      (() => {
        if (window !== window.top) return;
        const cap = window.Capacitor;
        if (!cap || typeof cap.toNative !== 'function') return;
        const original = cap.toNative;
        const activationId = %s;
        cap.toNative = function(plugin, method, options, callback) {
          if (plugin === 'OtaKit' && method === 'notifyAppReady') {
            options = Object.assign({}, options, { _otakitActivationId: activationId });
          }
          return original.call(this, plugin, method, options, callback);
        };
      })();
      """,
      JSONObject.quote(activationId)
    );
  }

  void install(WebView webView, String appUrl, String activationId) {
    requireSupported();
    // Register before removing the old script so registration failure preserves it.
    ScriptHandler replacement = WebViewCompat.addDocumentStartJavaScript(
      webView,
      source(activationId),
      Collections.singleton(origin(appUrl))
    );
    if (script != null) script.remove();
    script = replacement;
  }
}

File: packages/capacitor-plugin/ios/Sources/UpdaterPlugin/DocumentReadyBridge.swift
import Foundation
import WebKit

/// Captures the activation in the document's bridge before application JavaScript runs.
final class DocumentReadyBridge {
  private var script: WKUserScript?

  static func source(activationId: String) throws -> String {
    let quotedId = String(data: try JSONEncoder().encode(activationId), encoding: .utf8)!
    return """
    (() => {
      if (window !== window.top) return;
      const cap = window.Capacitor;
      if (!cap || typeof cap.toNative !== 'function') return;
      const original = cap.toNative;
      const activationId = \(quotedId);
      cap.toNative = function(plugin, method, options, callback) {
        if (plugin === 'OtaKit' && method === 'notifyAppReady') {
          options = Object.assign({}, options, { _otakitActivationId: activationId });
        }
        return original.call(this, plugin, method, options, callback);
      };
    })();
    """
  }

  func install(on webView: WKWebView, activationId: String) throws {
    precondition(Thread.isMainThread)
    let controller = webView.configuration.userContentController
    let replacement = WKUserScript(
      source: try Self.source(activationId: activationId),
      injectionTime: .atDocumentStart, forMainFrameOnly: true
    )
    // WebKit has no individual removal API. Preserve every other plugin's scripts and order.
    if let script {
      let retained = controller.userScripts.filter { $0 !== script }
      controller.removeAllUserScripts()
      for retainedScript in retained { controller.addUserScript(retainedScript) }
    }
    controller.addUserScript(replacement)
    script = replacement
  }
}
